### PR TITLE
return complete snapshot info in snapshots list

### PR DESCRIPTION
**Pull request by Madhu-1**
during the snapshot, list returns the complete snapshots info along with the parent volume name.
 if we call snapshot list we should be able to get complete snapshot information like listing volumes.

Signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

output:
```
[
    {
        "parentname": "test12",
        "snaps": [
            {
                "snapinfo": {
                    "id": "3b3c3951-78d4-4048-a794-5b54d7dc524a",
                    "name": "t11",
                    "type": "Distribute",
                    "transport": "tcp",
                    "distribute-count": 1,
                    "replica-count": 1,
                    "options": {
                        "bitrot-stub.bitrot": "off",
                        "feature.deem-statfs": "off",
                        "features.inode-quota": "off",
                        "features.quota": "off",
                        "features.quota-deem-statfs": "off",
                        "features.read-only": "on",
                        "features.uss": "off",
                        "replicate.self-heal-daemon": "off"
                    },
                    "state": "Started",
                    "subvols": [
                        {
                            "name": "t11-distribute-0",
                            "type": "Distribute",
                            "bricks": [
                                {
                                    "id": "f8f35aa2-aec4-4d1b-b95d-542c8d09b482",
                                    "path": "/usr/local/var/run/glusterd2/snaps/t11/test12/subvol1/brick1",
                                    "volume-id": "3b3c3951-78d4-4048-a794-5b54d7dc524a",
                                    "volume-name": "t11",
                                    "peer-id": "19d4532e-7ca3-4578-8bb2-927253436366",
                                    "host": "dhcp43-106.lab.eng.blr.redhat.com",
                                    "type": "Brick"
                                }
                            ],
                            "replica-count": 1,
                            "arbiter-count": 0,
                            "disperse-count": 0
                        }
                    ],
                    "metadata": {},
                    "snap-list": []
                },
                "parentname": "test12",
                "description": ""
            },
            {
                "snapinfo": {
                    "id": "7147b5f4-f235-4268-9e3e-bfcb1daedfb2",
                    "name": "te",
                    "type": "Distribute",
                    "transport": "tcp",
                    "distribute-count": 1,
                    "replica-count": 1,
                    "options": {
                        "bitrot-stub.bitrot": "off",
                        "feature.deem-statfs": "off",
                        "features.inode-quota": "off",
                        "features.quota": "off",
                        "features.quota-deem-statfs": "off",
                        "features.read-only": "on",
                        "features.uss": "off",
                        "replicate.self-heal-daemon": "off"
                    },
                    "state": "Created",
                    "subvols": [
                        {
                            "name": "te-distribute-0",
                            "type": "Distribute",
                            "bricks": [
                                {
                                    "id": "c73bb69a-f0b9-40ea-ae8f-e34bddf104fe",
                                    "path": "/usr/local/var/run/glusterd2/snaps/te/test12/subvol1/brick1",
                                    "volume-id": "7147b5f4-f235-4268-9e3e-bfcb1daedfb2",
                                    "volume-name": "te",
                                    "peer-id": "19d4532e-7ca3-4578-8bb2-927253436366",
                                    "host": "dhcp43-106.lab.eng.blr.redhat.com",
                                    "type": "Brick"
                                }
                            ],
                            "replica-count": 1,
                            "arbiter-count": 0,
                            "disperse-count": 0
                        }
                    ],
                    "metadata": {},
                    "snap-list": []
                },
                "parentname": "test12",
                "description": ""
            }
        ]
    }
]
```

### CODE REVIEW

**Comment by Madhu-1** [comment]
@rafikc30 @aravindavk @prashanthpai please review

File: pkg/api/snapshot_resp.go [hunk 1]
```diff
@@ -12,8 +12,8 @@ type SnapInfo struct {
 
 //SnapList contains snapshot name of a volume
```
Comment (rishubhjain) [inline pkg/api/snapshot_resp.go]: Please update the comment as well.

File: glusterd2/commands/snapshot/snapshot-list.go [hunk 2]
```diff
@@ -36,19 +35,28 @@ func snapshotListHandler(w http.ResponseWriter, r *http.Request) {
 			snapName[s.ParentVolume] = append(snapName[s.ParentVolume], s.SnapVolinfo.Name)
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glusterd2/commands/snapshot/snapshot-list.go]: but we can avoid the code duplication if we make use of the function `createSnapshotListResp`

[the remark this answers is not in the document]
Comment (rafikc30) [reply glusterd2/commands/snapshot/snapshot-list.go]: the redundant code in `createSnapshotListResp`, ie ```snapInfo, err := snapshot.GetSnapshot(s)``` is very costly operation. assume you have 1000+ snapshots in the cluster, that will badly affect the performance.

[the remark this answers is not in the document]
Comment (rafikc30) [reply glusterd2/commands/snapshot/snapshot-list.go]: I would even suggest to use two response function one for all snapshot, and one for all snapshot of a volume

**Comment by rafikc30** [comment]
Thinking more wider, it will be difficult for a general user who just want to list the snapshot name. And moreover, it will be a costlier operation to fetch all the details for a volume.

Do you guys think it would be more convenient to define another endpoint like, /v1/snapshots/info or  /v1/snapshots/info?volume=<volname> ?

File: glusterd2/commands/snapshot/snapshot-list.go [hunk 3]
```diff
@@ -23,32 +23,45 @@ func snapshotListHandler(w http.ResponseWriter, r *http.Request) {
...
+	resp, err := createSnapshotListResp(snapName)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
 	restutils.SendHTTPResponse(ctx, w, http.StatusOK, resp)
 }
 
-func createSnapshotListResp(snaps map[string][]string) *api.SnapListResp {
+func createSnapshotListResp(volSnaps map[string][]api.SnapInfo) (*api.SnapListResp, error) {
```
Comment (rafikc30) [inline glusterd2/commands/snapshot/snapshot-list.go]: No need to return error here

**Comment by Madhu-1** [comment]
@rafikc30 i have done some more changes please review

**Pull request merged**